### PR TITLE
C# reference: Consolidate the iteration statements articles

### LINKED ISSUE #10748: Abridged Reference: Statements / iteration statements
This will cover the following statements:

- `do` / `while`
- `for`
- `foreach` /  `in`
- `while`

Sub-tasks are:

- [x] Create the new articles
- [x] Search for existing issues on content being migrated and consolidated. This is a good opportunity to fix those.
- [x] Redirect existing content to the new articles.
- [x] Remove old articles.
- [x] Update all links in the .NET Content

### PR DESCRIPTION
**Pull request by pkulikov**
Fixes #10748

### CODE REVIEW

**Comment by pkulikov** [comment]
Close/open to restart the build

File: docs/csharp/language-reference/statements/iteration-statements.md [hunk 1]
```diff
@@ -0,0 +1,163 @@
...
+
+Beginning with C# 7.3, if the enumerator's `Current` property returns a [reference return value](../keywords/ref.md#reference-return-values) (`ref T` where `T` is the type of a collection element), you can declare an iteration variable with the `ref` or `ref readonly` modifier, as the following example shows:
+
+:::code language="csharp" source="snippets/iteration-statements/ForeachStatement.cs" id="RefIterationVariable" :::
+
+If the `foreach` statement is applied to `null`, a <xref:System.NullReferenceException> is thrown. If the source collection of the `foreach` statement is empty, the body of the `foreach` statement isn't executed and skipped.
+
+### await foreach
+
+Beginning with C# 8.0, you can use the `await foreach` statement to consume an asynchronous stream of data, that is, the collection type that implements the <xref:System.Collections.Generic.IAsyncEnumerable%601> interface. Each iteration of the loop may be suspended while the next element is retrieved asynchronously. The following example shows how to use the `await foreach` statement:
```
Comment (BillWagner) [inline docs/csharp/language-reference/statements/iteration-statements.md:106]: Is it worth mentioning `IAsyncEnumerator`, and its relationship with `IEnumerator`: meaning a `Current` property and a `MoveNextAsync` method that returns a `Task<bool>`?

I think yes, but that's a very weakly held opinion. Thoughts @pkulikov

**Review by BillWagner — APPROVED** [review]
Wow! Thanks for all the work here @pkulikov 

I'm really happy to approve this.

I have one question that I'd like to resolve before merging. Let me know your thoughts, and then this is ready.

Thanks again for all this work. 💯

File: docs/csharp/language-reference/statements/iteration-statements.md (hunk 1 above)
Comment (pkulikov) [reply docs/csharp/language-reference/statements/iteration-statements.md:106]: @BillWagner do you mean we should mention the pattern-based compilation as we do for a synchronous foreach? I think that would be useful to be consistent with the beginning of the section

Comment (BillWagner) [reply docs/csharp/language-reference/statements/iteration-statements.md:106]: @pkulikov  Yes, that's what I mean. Thanks.

Comment (pkulikov) [reply docs/csharp/language-reference/statements/iteration-statements.md:106]: @BillWagner I've updated the article; please check the latest commit.

Comment (BillWagner) [reply docs/csharp/language-reference/statements/iteration-statements.md:106]: That's perfect @pkulikov 

I'll :shipit: as soon as the build finishes.

Thanks again!

Comment (pkulikov) [reply docs/csharp/language-reference/statements/iteration-statements.md:106]: @BillWagner thank you for the review and the suggestion for `await foreach`

**Pull request merged**